<html><head></head><body>
<div id="sbo-rt-content"><div id="_idContainer001">
<h1 class="hidden">Bioinformatics with Python Cookbook <em class="italic">Third Edition</em></h1>
<p class="hidden">Use modern Python libraries and applications to solve real-world computational biology problems</p>
<p class="hidden">Tiago Antao</p>
<p><img alt="" height="141" src="image/Packt_Logo_SuperSite_2022_Orange.png" width="556"/></p>
<p>BIRMINGHAM—MUMBAI</p>
<h1 id="_idParaDest-1"><a id="_idTextAnchor000"/>Bioinformatics with Python Cookbook <em class="italic">Thi</em><em class="italic">rd Edition</em></h1>
<p>Copyright © 2022 Packt Publishing</p>
<p><em class="italic">All rights reserved</em>. No part of this book may be reproduced, stored in a retrieval system, or transmitted in any form or by any means, without the prior written permission of the publisher, except in the case of brief quotations embedded in critical articles or reviews.</p>
<p>Every effort has been made in the preparation of this book to ensure the accuracy of the information presented. However, the information contained in this book is sold without warranty, either express or implied. Neither the author, nor Packt Publishing or its dealers and distributors, will be held liable for any damages caused or alleged to have been caused directly or indirectly by this book.</p>
<p>Packt Publishing has endeavored to provide trademark information about all of the companies and products mentioned in this book by the appropriate use of capitals. However, Packt Publishing cannot guarantee the accuracy of this information.</p>
<p><strong class="bold">Publishing Product Manager</strong>: Devika Battike</p>
<p><strong class="bold">Senior Editor</strong>: David Sugarman</p>
<p><strong class="bold">Content Development Editor</strong>: Joseph Sunil</p>
<p><strong class="bold">Technical Editor</strong>: Rahul Limbachiya</p>
<p><strong class="bold">Copy Editor</strong>: Safis Editing</p>
<p><strong class="bold">Project Coordinator</strong>: Farheen Fathima</p>
<p><strong class="bold">Proofreader</strong>: Safis Editing</p>
<p><strong class="bold">Indexer</strong>: Pratik Shirodkar</p>
<p><strong class="bold">Production Designer</strong>: Shankar Kalbhor</p>
<p><strong class="bold">Marketing Coordinator</strong>: Priyanka Mhatre</p>
<p>First published: June 2015</p>
<p>Second edition: November 2018</p>
<p>Third edition: September 2022</p>
<p>Production reference: 1090922</p>
<p>Published by Packt Publishing Ltd.</p>
<p>Livery Place</p>
<p>35 Livery Street</p>
<p>Birmingham</p>
<p>B3 2PB, UK.</p>
<p>ISBN 978-1-80323-642-1</p>
<p><a href="http://www.packt.com">www.packt.com</a></p>
<h1 id="_idParaDest-2"><a id="_idTextAnchor001"/>Contributors</h1>
<h1 id="_idParaDest-3"><a id="_idTextAnchor002"/>About the author</h1>
<p><strong class="bold">Tiago Antao</strong> is a bioinformatician who is currently working in the field of genomics. A former computer scientist, Tiago moved into computational biology with an MSc in bioinformatics from the Faculty of Sciences at the University of Porto, Portugal, and a PhD on the spread of drug-resistant malaria from the Liverpool School of Tropical Medicine, UK. Post his doctoral, Tiago worked with human datasets at the University of Cambridge, UK and with mosquito whole-genome sequencing data at the University of Oxford, UK, before helping to set up the bioinformatics infrastructure at the University of Montana, USA. He currently works as a data engineer in the biotechnology field in Boston, MA. He is one of the co-authors of Biopython, a major bioinformatics package written in Python.</p>
<h1 id="_idParaDest-4"><a id="_idTextAnchor003"/>About the reviewers</h1>
<p><strong class="bold">Urminder Singh</strong> is a bioinformatician, computer scientist, and developer of multiple open source bioinformatics tools. His educational background encompasses physics, computer science, and computational biology degrees, including a Ph.D. in bioinformatics from Iowa State University, USA. </p>
<p>His diverse research interests include novel gene evolution, precision medicine, sociogenomics, machine learning in medicine, and developing tools and algorithms for big heterogeneous data. You can visit him online at urmi-21.github.io.</p>
<p><strong class="bold">Tiffany Ho</strong> works as a bioinformatics associate at Embark Veterinary. She holds a BSc from the University of California, Davis in genetics and genomics, and an MPS from Cornell University in plant breeding and genetics.</p>
</div>
<div>
<div id="_idContainer002">
</div>
</div>
</div>

<div id="sbo-rt-content"><div class="Content" epub:type="toc" id="_idContainer003">
<h1>Table of Contents</h1>
<h2><a href="B17942_Preface_eBook.xhtml#_idTextAnchor004">Preface</a></h2>
<h2><a href="B17942_01.xhtml#_idTextAnchor020">1</a></h2>
<h2><a href="B17942_01.xhtml#_idTextAnchor021">Python and the Surrounding Software Ecology</a></h2>
<h3><a href="B17942_01.xhtml#_idTextAnchor022">Installing the required basic software with Anaconda</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor023">Getting ready</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor024">How to do it...</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor025">There’s more...</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor026">Installing the required software with Docker</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor027">Getting ready</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor028">How to do it...</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor029">See also</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor030">Interfacing with R via rpy2</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor031">Getting ready</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor032">How to do it...</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor033">There’s more...</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor034">See also</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor035">Performing R magic with Jupyter</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor036">Getting ready</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor037">How to do it...</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor038">There’s more...</a></h3>
<h3><a href="B17942_01.xhtml#_idTextAnchor039">See also</a></h3>
<h2><a href="B17942_02.xhtml#_idTextAnchor040">2</a></h2>
<h2><a href="B17942_02.xhtml#_idTextAnchor041">Getting to Know NumPy, pandas, Arrow, and Matplotlib</a></h2>
<h3><a href="B17942_02.xhtml#_idTextAnchor042">Using pandas to process vaccine-adverse events </a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor043">Getting ready</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor044">How to do it...</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor045">There’s more...</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor046">See also</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor047">Dealing with the pitfalls of joining pandas DataFrames</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor048">Getting ready</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor049">How to do it...</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor050">There’s more...</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor051">Reducing the memory usage of pandas DataFrames</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor052">Getting ready</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor053">How to do it…</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor054">See also</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor055">Accelerating pandas processing with Apache Arrow</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor056">Getting ready</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor057">How to do it...</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor058">There’s more...</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor059">Understanding NumPy as the engine behind Python data science and bioinformatics</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor060">Getting ready</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor061">How to do it…</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor062">See also</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor063">Introducing Matplotlib for chart generation</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor064">Getting ready</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor065">How to do it...</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor066">There’s more...</a></h3>
<h3><a href="B17942_02.xhtml#_idTextAnchor067">See also</a></h3>
<h2><a href="B17942_03.xhtml#_idTextAnchor068">3</a></h2>
<h2><a href="B17942_03.xhtml#_idTextAnchor069">Next-Generation Sequencing</a></h2>
<h3><a href="B17942_03.xhtml#_idTextAnchor070">Accessing GenBank and moving around NCBI databases</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor071">Getting ready</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor072">How to do it...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor073">There’s more...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor074">See also</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor075">Performing basic sequence analysis</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor076">Getting ready</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor077">How to do it...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor078">There’s more...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor079">See also</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor080">Working with modern sequence formats</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor081">Getting ready</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor082">How to do it...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor083">There’s more...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor084">See also</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor085">Working with alignment data</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor086">Getting ready</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor087">How to do it...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor088">There’s more...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor089">See also</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor090">Extracting data from VCF files</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor091">Getting ready</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor092">How to do it...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor093">There’s more...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor094">See also</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor095">Studying genome accessibility and filtering SNP data</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor096">Getting ready</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor097">How to do it...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor098">There’s more...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor099">See also</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor100">Processing NGS data with HTSeq</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor101">Getting ready</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor102">How to do it...</a></h3>
<h3><a href="B17942_03.xhtml#_idTextAnchor103">There’s more...</a></h3>
<h2><a href="B17942_04.xhtml#_idTextAnchor104">4</a></h2>
<h2><a href="B17942_04.xhtml#_idTextAnchor105">Advanced NGS Data Processing</a></h2>
<h3><a href="B17942_04.xhtml#_idTextAnchor106">Preparing a dataset for analysis</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor107">Getting ready</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor108">How to do it…</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor109">Using Mendelian error information for quality control</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor110">How to do it…</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor111">There’s more…</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor112">Exploring the data with standard statistics</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor113">How to do it…</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor114">There’s more…</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor115">Finding genomic features from sequencing annotations</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor116">How to do it…</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor117">There’s more…</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor118">Doing metagenomics with QIIME 2 Python API</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor119">Getting ready</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor120">How to do it...</a></h3>
<h3><a href="B17942_04.xhtml#_idTextAnchor121">There’s more...</a></h3>
<h2><a href="B17942_05.xhtml#_idTextAnchor122">5</a></h2>
<h2><a href="B17942_05.xhtml#_idTextAnchor123">Working with Genomes</a></h2>
<h3><a href="B17942_05.xhtml#_idTextAnchor124">Technical requirements</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor125">Working with high-quality reference genomes</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor126">Getting ready</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor127">How to do it...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor128">There’s more...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor129">See also</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor130">Dealing with low-quality genome references</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor131">Getting ready</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor132">How to do it...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor133">There’s more...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor134">See also</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor135">Traversing genome annotations</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor136">Getting ready</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor137">How to do it...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor138">There’s more...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor139">See also</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor140">Extracting genes from a reference using annotations</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor141">Getting ready</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor142">How to do it...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor143">There’s more...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor144">See also</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor145">Finding orthologues with the Ensembl REST API</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor146">Getting ready</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor147">How to do it...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor148">There’s more...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor149">Retrieving gene ontology information from Ensembl</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor150">Getting ready</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor151">How to do it...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor152">There’s more...</a></h3>
<h3><a href="B17942_05.xhtml#_idTextAnchor153">See also</a></h3>
<h2><a href="B17942_06.xhtml#_idTextAnchor154">6</a></h2>
<h2><a href="B17942_06.xhtml#_idTextAnchor155">Population Genetics</a></h2>
<h3><a href="B17942_06.xhtml#_idTextAnchor156">Managing datasets with PLINK</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor157">Getting ready</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor158">How to do it...</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor159">There’s more...</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor160">See also</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor161">Using sgkit for population genetics analysis with xarray</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor162">Getting ready</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor163">How to do it...</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor164">There’s more...</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor165">Exploring a dataset with sgkit</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor166">Getting ready</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor167">How to do it...</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor168">There’s more...</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor169">See also</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor170">Analyzing population structure</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor171">Getting ready</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor172">How to do it...</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor173">See also</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor174">Performing a PCA</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor175">Getting ready</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor176">How to do it...</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor177">There’s more...</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor178">See also</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor179">Investigating population structure with admixture</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor180">Getting ready</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor181">How to do it...</a></h3>
<h3><a href="B17942_06.xhtml#_idTextAnchor182">There’s more...</a></h3>
<h2><a href="B17942_07.xhtml#_idTextAnchor183">7</a></h2>
<h2><a href="B17942_07.xhtml#_idTextAnchor184">Phylogenetics</a></h2>
<h3><a href="B17942_07.xhtml#_idTextAnchor185">Preparing a dataset for phylogenetic analysis</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor186">Getting ready</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor187">How to do it...</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor188">There’s more...</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor189">See also</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor190">Aligning genetic and genomic data</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor191">Getting ready</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor192">How to do it...</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor193">Comparing sequences</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor194">Getting ready</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor195">How to do it...</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor196">There’s more...</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor197">Reconstructing phylogenetic trees</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor198">Getting ready</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor199">How to do it...</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor200">There’s more...</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor201">Playing recursively with trees</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor202">Getting ready</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor203">How to do it...</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor204">There’s more...</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor205">Visualizing phylogenetic data</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor206">Getting ready</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor207">How to do it...</a></h3>
<h3><a href="B17942_07.xhtml#_idTextAnchor208">There’s more...</a></h3>
<h2><a href="B17942_08.xhtml#_idTextAnchor209">8</a></h2>
<h2><a href="B17942_08.xhtml#_idTextAnchor210">Using the Protein Data Bank</a></h2>
<h3><a href="B17942_08.xhtml#_idTextAnchor211">Finding a protein in multiple databases</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor212">Getting ready</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor213">How to do it...</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor214">There’s more</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor215">Introducing Bio.PDB</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor216">Getting ready</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor217">How to do it...</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor218">There’s more</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor219">Extracting more information from a PDB file</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor220">Getting ready</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor221">How to do it...</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor222">Computing molecular distances on a PDB file</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor223">Getting ready</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor224">How to do it...</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor225">Performing geometric operations</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor226">Getting ready</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor227">How to do it...</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor228">There’s more</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor229">Animating with PyMOL</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor230">Getting ready</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor231">How to do it...</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor232">There’s more</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor233">Parsing mmCIF files using Biopython</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor234">Getting ready</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor235">How to do it...</a></h3>
<h3><a href="B17942_08.xhtml#_idTextAnchor236">There’s more</a></h3>
<h2><a href="B17942_09.xhtml#_idTextAnchor237">9</a></h2>
<h2><a href="B17942_09.xhtml#_idTextAnchor238">Bioinformatics Pipelines</a></h2>
<h3><a href="B17942_09.xhtml#_idTextAnchor239">Introducing Galaxy servers</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor240">Getting ready</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor241">How to do it…</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor242">There’s more</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor243">Accessing Galaxy using the API</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor244">Getting ready</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor245">How to do it…</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor246">Deploying a variant analysis pipeline with Snakemake</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor248">Getting ready</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor249">How to do it…</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor250">There’s more</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor251">Deploying a variant analysis pipeline with Nextflow</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor252">Getting ready</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor253">How to do it…</a></h3>
<h3><a href="B17942_09.xhtml#_idTextAnchor254">There’s more</a></h3>
<h2><a href="B17942_10.xhtml#_idTextAnchor255">10</a></h2>
<h2><a href="B17942_10.xhtml#_idTextAnchor256">Machine Learning for Bioinformatics</a></h2>
<h3><a href="B17942_10.xhtml#_idTextAnchor257">Introducing scikit-learn with a PCA example</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor258">Getting ready</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor259">How to do it...</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor260">There’s more...</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor261">Using clustering over PCA to classify samples</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor262">Getting ready</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor263">How to do it...</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor264">There’s more...</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor265">Exploring breast cancer traits using Decision Trees</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor266">Getting ready</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor267">How to do it...</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor268">Predicting breast cancer outcomes using Random Forests </a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor269">Getting ready</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor270">How to do it…</a></h3>
<h3><a href="B17942_10.xhtml#_idTextAnchor271">There’s more...</a></h3>
<h2><a href="B17942_11.xhtml#_idTextAnchor272">11</a></h2>
<h2><a href="B17942_11.xhtml#_idTextAnchor273">Parallel Processing with Dask and Zarr</a></h2>
<h3><a href="B17942_11.xhtml#_idTextAnchor274">Reading genomics data with Zarr </a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor275">Getting ready</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor276">How to do it...</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor277">There’s more...</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor278">See also</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor279">Parallel processing of data using Python multiprocessing</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor280">Getting ready</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor281">How to do it...</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor282">There’s more...</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor283">See also</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor284">Using Dask to process genomic data based on NumPy arrays</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor285">Getting ready</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor286">How to do it...</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor287">There’s more...</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor288">See also</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor289">Scheduling tasks with dask.distributed</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor290">Getting ready</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor291">How to do it...</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor292">There’s more...</a></h3>
<h3><a href="B17942_11.xhtml#_idTextAnchor293">See also</a></h3>
<h2><a href="B17942_12.xhtml#_idTextAnchor294">12</a></h2>
<h2><a href="B17942_12.xhtml#_idTextAnchor295">Functional Programming for Bioinformatics</a></h2>
<h3><a href="B17942_12.xhtml#_idTextAnchor296">Understanding pure functions </a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor297">Getting ready</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor298">How to do it...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor299">There’s more...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor300">Understanding immutability</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor301">Getting ready</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor302">How to do it...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor303">There’s more...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor304">Avoiding mutability as a robust development pattern</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor305">Getting ready</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor306">How to do it...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor307">There’s more...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor308">Using lazy programming for pipelining</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor309">Getting ready</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor310">How to do it...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor311">There’s more...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor312">The limits of recursion with Python</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor313">Getting ready</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor314">How to do it...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor315">There’s more...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor316">A showcase of Python’s functools module</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor317">Getting ready</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor318">How to do it...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor319">There’s more...</a></h3>
<h3><a href="B17942_12.xhtml#_idTextAnchor320">See also...</a></h3>
<h2><a href="B17942_Index.xhtml#_idTextAnchor321">Index</a></h2>
<h2><a href="B17942_BM_eBook.xhtml#_idTextAnchor323">Other Books You May Enjoy</a></h2>
</div>
</div></body></html>