<html><head></head><body>
<div id="sbo-rt-content"><div id="_idContainer004">
<h1 id="_idParaDest-5"><a id="_idTextAnchor004"/>Preface</h1>
<p>Bioinformatics is an active research field that uses a range of simple-to-advanced computations to extract valuable information from biological data, and this book will show you how to manage these tasks using Python.</p>
<p>This updated edition of the Bioinformatics with Python Cookbook begins with a quick overview of the various tools and libraries in the Python ecosystem that will help you convert, analyze, and visualize biological datasets. As you advance through the chapters, you’ll cover key techniques for next-generation sequencing, single-cell analysis, genomics, metagenomics, population genetics, phylogenetics, and proteomics with the help of real-world examples. You’ll learn how to work with important pipeline systems, such as Galaxy servers and Snakemake, and understand the various modules in Python for functional and asynchronous programming. This book will also help you explore topics such as SNP discovery using statistical approaches under high-performance computing frameworks, including Dask and Spark, and the application of machine learning algorithms to bioinformatics.</p>
<p>By the end of this bioinformatics Python book, you’ll be equipped with the knowledge to implement the latest programming techniques and frameworks, empowering you to deal with bioinformatics data on every kind of scale.</p>
<h1 id="_idParaDest-6"><a id="_idTextAnchor005"/>Who this book is for</h1>
<p>This book is for bioinformatics analysts, data scientists, computational biologists, researchers, and Python developers who want to address intermediate-to-advanced biological and bioinformatics problems. Working knowledge of the Python programming language is expected. Basic knowledge of biology would be helpful.</p>
<h1 id="_idParaDest-7"><a id="_idTextAnchor006"/>What this book covers</h1>
<p><a href="B17942_01.xhtml#_idTextAnchor020"><em class="italic">Chapter 1</em></a><em class="italic">, Python and the Surrounding Software Ecology</em>, tells you how to set up a modern bioinformatics environment with Python. This chapter discusses how to deploy software using Docker, interface with R, and interact with the Jupyter Notebooks.</p>
<p><a href="B17942_02.xhtml#_idTextAnchor040"><em class="italic">Chapter 2</em></a><em class="italic">, Getting to Know NumPy, pandas, Arrow, and Matplotlib</em>, introduces the fundamental Python libraries for data science: NumPy for array and matrix processing; Pandas for table-based data manipulation; Arrow to optimize Pandas processing and Matplotlib for charting.</p>
<p><a href="B17942_03.xhtml#_idTextAnchor068"><em class="italic">Chapter 3</em></a><em class="italic">, Next-Generation Sequencing</em>, provides concrete solutions to deal with next-generation sequencing data. This chapter teaches you how to deal with large FASTQ, BAM, and VCF files. It also discusses data filtering. </p>
<p><a href="B17942_04.xhtml#_idTextAnchor104"><em class="italic">Chapter 4</em></a><em class="italic">, Advanced NGS Processing</em>, covers advanced programming techniques to filter NGS data. This includes the use of mendelian datasets that are then analyzed by standard statistics. We also introduce metagenomic analysis</p>
<p><a href="B17942_05.xhtml#_idTextAnchor122"><em class="italic">Chapter 5</em></a><em class="italic">, Working with Genomes</em>, not only deals with high-quality references—such as the human genome—but also discusses how to analyze other low-quality references typical in nonmodel species. It introduces GFF processing, teaches you to analyze genomic feature information, and discusses how to use gene ontologies. </p>
<p><a href="B17942_06.xhtml#_idTextAnchor154"><em class="italic">Chapter 6</em></a><em class="italic">, Population Genetics</em>, describes how to perform population genetics analysis of empirical datasets. For example, in Python, we could perform Principal Components Analysis, computer FST, or structure/admixture plots. </p>
<p><a href="B17942_07.xhtml#_idTextAnchor183"><em class="italic">Chapter 7</em></a><em class="italic">, Phylogenetics</em>, uses complete sequences of recently sequenced Ebola viruses to perform real phylogenetic analysis, which includes tree reconstruction and sequence comparisons. This chapter discusses recursive algorithms to process tree-like structures.</p>
<p><a href="B17942_08.xhtml#_idTextAnchor209"><em class="italic">Chapter 8</em></a><em class="italic">, Using the Protein Data Bank</em>, focuses on processing PDB files, for example, performing the geometric analysis of proteins. This chapter takes a look at protein visualization.</p>
<p><a href="B17942_09.xhtml#_idTextAnchor237"><em class="italic">Chapter 9</em></a><em class="italic">, Bioinformatics Pipelines</em>, introduces two types of pipelines. The first type of pipeline is Python-based Galaxy, a widely used system with a web interface targeting mostly non-programming users although bioinformaticians might still have to interact with it programmatically. The second type will be based on snakemake and nextflow, a type of pipeline that targets programmers.</p>
<p><a href="B17942_10.xhtml#_idTextAnchor255"><em class="italic">Chapter 10</em></a><em class="italic">, Machine Learning for Bioinformatics</em>, introduces machine learning using an intuitive approach to deal with computational biology problems. The chapter covers Principal Components Analysis, Clustering, Decision Trees, and Random Forests.</p>
<p><a href="B17942_11.xhtml#_idTextAnchor272"><em class="italic">Chapter 11</em></a><em class="italic">, Parallel Processing with Dask and Zarr</em>, introduces techniques to deal with very large datasets and computationally intensive algorithms. The chapter will explain how to use parallel computation across many computers (cluster or cloud). We will also discuss the efficient storage of biological data.</p>
<p><a href="B17942_12.xhtml#_idTextAnchor294"><em class="italic">Chapter 12</em></a><em class="italic">, Functional Programming for Bioinformatics</em>, introduces functional programming which permits the development of more sophisticated Python programs that, through lazy programming and immutability are easier to deploy in parallel environments with complex algorithms</p>
<h1 id="_idParaDest-8"><a id="_idTextAnchor007"/>To get the most out of this book</h1>
<table class="No-Table-Style _idGenTablePara-1" id="table001">
<colgroup>
<col/>
<col/>
</colgroup>
<tbody>
<tr class="No-Table-Style">
<td class="No-Table-Style">
<p><strong class="bold">Software/Hardware covered in the book</strong></p>
</td>
<td class="No-Table-Style">
<p><strong class="bold">OS Requirements</strong></p>
</td>
</tr>
<tr class="No-Table-Style">
<td class="No-Table-Style">
<p>Python 3.9</p>
</td>
<td class="No-Table-Style">
<p>Windows, Mac OS X, and Linux (Preferred)</p>
</td>
</tr>
<tr class="No-Table-Style">
<td class="No-Table-Style">
<p>Numpy, Pandas, Matplolib</p>
</td>
<td class="No-Table-Style"/>
</tr>
<tr class="No-Table-Style">
<td class="No-Table-Style">
<p>Biopython</p>
</td>
<td class="No-Table-Style"/>
</tr>
<tr class="No-Table-Style">
<td class="No-Table-Style">
<p>Dask, zarr, scikit-learn</p>
</td>
<td class="No-Table-Style"/>
</tr>
</tbody>
</table>
<p><strong class="bold">If you are using the digital version of this book, we advise you to type the code yourself or access the code via the GitHub repository (link available in the next section). Doing so will help you avoid any potential errors related to the copying and pasting of code.</strong></p>
<h2 id="_idParaDest-9"><a id="_idTextAnchor008"/>Download the example code files</h2>
<p>You can download the example code files for this book from GitHub at <a href="https://github.com/PacktPublishing/Bioinformatics-with-Python-Cookbook-third-edition">https://github.com/PacktPublishing/Bioinformatics-with-Python-Cookbook-third-edition</a>. In case there’s an update to the code, it will be updated on the existing GitHub repository.</p>
<p>We also have other code bundles from our rich catalog of books and videos available at <a href="https://github.com/PacktPublishing/">https://github.com/PacktPublishing/</a>. Check them out!</p>
<h1 id="_idParaDest-10"><a id="_idTextAnchor009"/>Download the color images</h1>
<p>We also provide a PDF file that has color images of the screenshots/diagrams used in this book. You can download it here: <a href="https://packt.link/3KQQO">https://packt.link/3KQQO</a>.</p>
<h1 id="_idParaDest-11"><a id="_idTextAnchor010"/>Conventions used</h1>
<p>There are a number of text conventions used throughout this book.</p>
<p><strong class="source-inline">Code in text</strong>: Indicates code words in text, database table names, folder names, filenames, file extensions, pathnames, dummy URLs, user input, and Twitter handles. Here is an example: “<strong class="source-inline">call_genotype</strong> has a shape of 56,241x1,1198,2, that is it is dimensioned variants, samples, <strong class="source-inline">ploidy</strong>.”</p>
<p>A block of code is set as follows:</p>
<pre class="source-code">
from Bio import SeqIO
genome_name = 'PlasmoDB-9.3_Pfalciparum3D7_Genome.fasta'
recs = SeqIO.parse(genome_name, 'fasta')
for rec in recs:
    print(rec.description)</pre>
<p>When we wish to draw your attention to a particular part of a code block, the relevant lines or items are set in bold:</p>
<pre class="source-code">
<strong class="bold">AgamP4_2L | organism=Anopheles_gambiae_PEST | version=AgamP4 | length=49364325 | SO=chromosome</strong>
<strong class="bold">AgamP4_2R | organism=Anopheles_gambiae_PEST | version=AgamP4 | length=61545105 | SO=chromosome</strong></pre>
<p><strong class="bold">Bold</strong>: Indicates a new term, an important word, or words that you see onscreen. For example, words in menus or dialog boxes appear in the text like this. Here is an example: “For the <strong class="bold">Chunk</strong> column, see <a href="B17942_11.xhtml#_idTextAnchor272"><em class="italic">Chapter 11</em></a> – but you can safely ignore it for now.”</p>
<p class="callout-heading">Tips or important notes</p>
<p class="callout">Appear like this.</p>
<h1 id="_idParaDest-12"><a id="_idTextAnchor011"/>Sections</h1>
<p>In this book, you will find several headings that appear frequently (<em class="italic">Getting ready</em>, <em class="italic">How to do it...</em>, <em class="italic">How it works...</em>, <em class="italic">There’s more...</em>, and <em class="italic">See also</em>).</p>
<p>To give clear instructions on how to complete a recipe, use these sections as follows:</p>
<h2 id="_idParaDest-13"><a id="_idTextAnchor012"/>Getting ready</h2>
<p>This section tells you what to expect in the recipe and describes how to set up any software or any preliminary settings required for the recipe.</p>
<h2 id="_idParaDest-14"><a id="_idTextAnchor013"/>How to do it…</h2>
<p>This section contains the steps required to follow the recipe.</p>
<h2 id="_idParaDest-15"><a id="_idTextAnchor014"/>How it works…</h2>
<p>This section usually consists of a detailed explanation of what happened in the previous section.</p>
<h2 id="_idParaDest-16"><a id="_idTextAnchor015"/>There’s more…</h2>
<p>This section consists of additional information about the recipe in order to make you more knowledgeable about the recipe.</p>
<h2 id="_idParaDest-17"><a id="_idTextAnchor016"/>See also</h2>
<p>This section provides helpful links to other useful information for the recipe.</p>
<h1 id="_idParaDest-18"><a id="_idTextAnchor017"/>Get in touch</h1>
<p>Feedback from our readers is always welcome.</p>
<p><strong class="bold">General feedback</strong>: If you have questions about any aspect of this book, mention the book title in the subject of your message and email us at <a href="https://customercare@packtpub.com">customercare@packtpub.com</a>.</p>
<p><strong class="bold">Errata</strong>: Although we have taken every care to ensure the accuracy of our content, mistakes do happen. If you have found a mistake in this book, we would be grateful if you would report this to us. Please visit <a href="https://www.packtpub.com/support/errata">www.packtpub.com/support/errata</a>, selecting your book, clicking on the Errata Submission Form link, and entering the details.</p>
<p><strong class="bold">Piracy</strong>: If you come across any illegal copies of our works in any form on the Internet, we would be grateful if you would provide us with the location address or website name. Please contact us at <a href="https://copyright@packt.com">copyright@packt.com</a> with a link to the material.</p>
<p><strong class="bold">If you are interested in becoming an author</strong>: If there is a topic that you have expertise in and you are interested in either writing or contributing to a book, please visit <a href="https://authors.packtpub.com">authors.packtpub.com</a>.</p>
<h1 id="_idParaDest-19"><a id="_idTextAnchor018"/>Reviews</h1>
<p>Please leave a review. Once you have read and used this book, why not leave a review on the site that you purchased it from? Potential readers can then see and use your unbiased opinion to make purchase decisions, we at Packt can understand what you think about our products, and our authors can see your feedback on their book. Thank you!</p>
<p>For more information about Packt, please visit <a href="https://packt.com">packt.com</a>.</p>
<h1 id="_idParaDest-20"><a id="_idTextAnchor019"/>Share Your Thoughts</h1>
<p>Once you’ve read <em class="italic">Bioinformatics with Python Cookbook</em>, we’d love to hear your thoughts! Please <a href="">click here to go straight to the Amazon review page</a> for this book and share your feedback.</p>
<p>Your review is important to us and the tech community and will help us make sure we’re delivering excellent quality content.</p>
</div>
</div></body></html>